contracts/GasStationExact.sol

```diff
@@ -0,0 +1,287 @@
+    for (uint256 idx = 0; idx < oldWatchList.length; idx++) {
+      s_targets[oldWatchList[idx]].isActive = false;
+    }
+    for (uint256 idx = 0; idx < addresses.length; idx++) {
+      if (s_targets[addresses[idx]].isActive) {
+        revert DuplicateAddress(addresses[idx]);
+      }
+      if (addresses[idx] == address(0)) {
+        revert InvalidWatchList();
+      }
+      if (topUpAmountsWei[idx] == 0) {
+        revert InvalidWatchList();
+      }
+      s_targets[addresses[idx]] = Target({
+        isActive: true,
+        minBalanceWei: minBalancesWei[idx],
+        minTopUpAmountWei: topUpAmountsWei[idx],
+        lastTopUpTimestamp: 0
+      });
+    }
+    s_watchList = addresses;
+  }
+
+  /**
+   * @notice Gets a list of addresses that are under funded
+   * @return list of addresses that are underfunded
+   */
+  function getUnderfundedAddresses() public view returns (address[] memory) {
+    address[] memory watchList = s_watchList;
+    address[] memory needsFunding = new address[](watchList.length);
+    uint256 count = 0;
+    uint256 minWaitPeriod = s_minWaitPeriodSeconds;
+    uint256 balance = address(this).balance;
+    Target memory target;
+    for (uint256 idx = 0; idx < watchList.length; idx++) {
+      target = s_targets[watchList[idx]];
+      if (watchList[idx].balance < target.minBalanceWei) {
+        uint256 delta = target.minBalanceWei - watchList[idx].balance;
+        if (
+          target.lastTopUpTimestamp + minWaitPeriod <= block.timestamp &&
+          balance >= delta &&
+          watchList[idx].balance <= target.minBalanceWei - target.minTopUpAmountWei
+        ) {
+          needsFunding[count] = watchList[idx];
+          count++;
+          balance -= delta;
+        }
+      }
+    }
+    if (count != watchList.length) {
+      assembly {
+        mstore(needsFunding, count)
+      }
+    }
+    return needsFunding;
+  }
+
+  /**
+   * @notice Send funds to the addresses provided
+   * @param needsFunding the list of addresses to fund (addresses must be pre-approved)
+   */
+  function topUpExact(address[] memory needsFunding) public whenNotPaused {
+    uint256 minWaitPeriodSeconds = s_minWaitPeriodSeconds;
+    Target memory target;
+    for (uint256 idx = 0; idx < needsFunding.length; idx++) {
+      target = s_targets[needsFunding[idx]];
+      uint256 delta = target.minBalanceWei - needsFunding[idx].balance;
+      if (
+        target.isActive &&
+        target.lastTopUpTimestamp + minWaitPeriodSeconds <= block.timestamp &&
+        delta >= target.minTopUpAmountWei
+      ) {
+        bool success = payable(needsFunding[idx]).send(delta);
+        if (success) {
+          s_targets[needsFunding[idx]].lastTopUpTimestamp = uint56(block.timestamp);
+          emit TopUpSucceeded(needsFunding[idx]);
+        } else {
+          emit TopUpFailed(needsFunding[idx]);
+        }
+      }
+      if (gasleft() < MIN_GAS_FOR_TRANSFER) {
+        return;
+      }
```

Comment:
Interesting. Just fyi, I know this is coming from CL but from what I know, hardcoding gas is usually not a good idea since opcode gas might change

Reply:
i understand it as a mere estimation no? ie is the loop not getting too big.

would there be a way to better define `MIN_GAS_FOR_TRANSFER`?

Reply:
Yep, it's not a big deal. Just that if the transfer gas cost increases above this minimum then txs might start reverting with oog. Not sure if there's a better way tbh